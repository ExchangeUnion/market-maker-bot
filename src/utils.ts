import moment from 'moment';
import { Observable, Subject } from 'rxjs';
import { take, tap } from 'rxjs/operators';
import { curry } from 'ramda';

/** Get the current date in the given dateFormat, if not provided formats with `YYYY-MM-DD hh:mm:ss.sss`.
 */
export const getTsString = (dateFormat?: string): string =>
  moment().format(dateFormat || 'YYYY-MM-DD hh:mm:ss.sss');

const SATOSHIS_PER_COIN = 10 ** 8;

/** Returns a number of satoshis as a string representation of coins with up to 8 decimal places. */
export const satsToCoinsStr = (satsQuantity: number): string => {
  return (satsQuantity / SATOSHIS_PER_COIN).toFixed(8).replace(/\.?0+$/, '');
};

/** Returns a number of coins as an integer number of satoshis. */
export const coinsToSats = (coinsQuantity: number): number => {
  return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
};

export const getStartShutdown$ = (): Observable<unknown> => {
  const shutdown$ = new Subject();
  process.on('SIGINT', () => shutdown$.next());
  process.on('SIGTERM', () => shutdown$.next());
  return shutdown$.asObservable().pipe(take(1));
};

const debugObservable = (prefix: string, source: Observable<any>) => {
  return source.pipe(
    tap({
      next: v => console.log(`${prefix} next: ${v}`),
      error: e => console.log(`${prefix} error: ${e}`),
      complete: () => console.log(`${prefix} complete`),
    })
  );
};

const debugLog = curry(debugObservable);

export { debugLog };
