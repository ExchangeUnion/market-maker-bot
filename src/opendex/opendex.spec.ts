import { TestScheduler } from 'rxjs/testing';
import {
  getOpenDEXassets$,
  xudBalanceToExchangeAssetAllocation,
  logAssetBalance,
} from './opendex';
import { Observable } from 'rxjs';
import { GetBalanceResponse } from '../broker/opendex/proto/xudrpc_pb';
import { BigNumber } from 'bignumber.js';
import { XudClient } from '../broker/opendex/proto/xudrpc_grpc_pb';
import { testConfig, getLoggers } from '../../test/utils';
import { Logger } from '../logger';

describe('OpenDEX', () => {
  let testScheduler: TestScheduler;

  beforeEach(() => {
    testScheduler = new TestScheduler((actual, expected) => {
      expect(actual).toEqual(expected);
    });
  });

  test('xudBalanceToExchangeAssetAllocation', () => {
    const ethBalance = {
      getChannelBalance: () => 10000000
    };
    const btcBalance = {
      getChannelBalance: () => 20000000
    };
    const balancesMap = new Map();
    const baseAsset = 'ETH';
    const quoteAsset = 'BTC';
    balancesMap.set(baseAsset, ethBalance);
    balancesMap.set(quoteAsset, btcBalance);
    const getBalanceResponse = {
      getBalancesMap: () => balancesMap,
    } as unknown as GetBalanceResponse;
    expect(
      xudBalanceToExchangeAssetAllocation({
        baseAsset,
        quoteAsset,
        balanceResponse: getBalanceResponse,
      })
    ).toEqual({
      baseAssetBalance:
        new BigNumber(ethBalance.getChannelBalance()),
      quoteAssetBalance:
        new BigNumber(btcBalance.getChannelBalance()),
    });
  });

  test('xudBalanceToExchangeAssetAllocation error baseAssetBalance', () => {
    const balancesMap = new Map();
    const getBalanceResponse = {
      getBalancesMap: () => balancesMap,
    } as unknown as GetBalanceResponse;
    expect(() => {
      xudBalanceToExchangeAssetAllocation({
        baseAsset: 'ETH',
        quoteAsset: 'BTC',
        balanceResponse: getBalanceResponse,
      })
    }).toThrowErrorMatchingSnapshot();
  });

  test('getOpenDexAssets$', () => {
    testScheduler.run(helpers => {
      const { cold, expectObservable } = helpers;
      const inputEvents = {
        xudBalance$:   '500ms a',
        xudClient$:    '500ms a',
      };
      const expected = '1s (a|)'
      const xudBalanceToExchangeAssetAllocation =
        ({ balanceResponse }: any) => balanceResponse;
      const getXudBalance$ = () => {
        return cold(inputEvents.xudBalance$) as unknown as Observable<GetBalanceResponse>;
      };
      const getXudClient$ = () => {
        return cold(inputEvents.xudClient$) as unknown as Observable<XudClient>;
      };
      const logBalance = () => {};
      const openDEXassets$ = getOpenDEXassets$({
        logBalance,
        config: testConfig(),
        logger: getLoggers().global,
        xudClient$: getXudClient$,
        xudBalance$: getXudBalance$,
        xudBalanceToExchangeAssetAllocation,
      });
      expectObservable(openDEXassets$).toBe(expected);
    });
  });

  describe('logAssetBalance', () => {
    const logger = getLoggers().global as unknown as Logger;
    logger.trace = jest.fn();
    const baseAssetBalance = '1.23456789';
    const quoteAssetBalance = '9.87654321';
    const assetBalance = {
      baseAssetBalance: new BigNumber(baseAssetBalance),
      quoteAssetBalance: new BigNumber(quoteAssetBalance),
    };
    logAssetBalance({
      logger,
      assetBalance,
    });
    expect(logger.trace).toHaveBeenCalledTimes(1);
    expect(logger.trace).toHaveBeenCalledWith(
      expect.stringContaining(baseAssetBalance)
    );
    expect(logger.trace).toHaveBeenCalledWith(
      expect.stringContaining(quoteAssetBalance)
    );
  });

});
