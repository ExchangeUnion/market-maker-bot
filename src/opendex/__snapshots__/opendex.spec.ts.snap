// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenDEX xudBalanceToExchangeAssetAllocation error baseAssetBalance 1`] = `"OpenDEX balance did not include balance for base asset ETH or quote asset BTC."`;
